### ISSUE TITLE
Can taskflow achieve functions like CUDA Stream and Event?

### ISSUE THREAD

**Issue by ccsuzzh** [taskflow/taskflow#457]
**Can taskflow achieve functions like CUDA Stream and Event?**
Taskflow is a great job! I want to use it in my project.  The upper layer of the framework provides such APIs, like device.cuda.Stream(), device.cuda.Event(), device.cuda.synchronize() etc, for users to write multi-stream programs, but there is no Stream component in our device, so can I use taskflow to achieve functions like CUDA Stream and Event?

**Comment by tsung-wei-huang** [comment]
@ccsuzzh , thank you for the question. Could you please be a bit more specific about your need (perhaps with some examples)? Taskflow itself doesn't offer any stream facility but let you define your applications in a task graph. Our scheduler will schedule that graph with dynamic load balance. You can always define your own GPU kernels in a task or multiple tasks with dependencies.

**Comment by ccsuzzh** [comment]
Thanks for your answer, I plan to implement a parallel scheduling mode similar to CUDA-based stream&event on a custom platform. Our current application scenario is that for an A→(B and C)→D instance, task B can be defined by specifying its dependencies: task A (e.g. B.succed (A)) and executing it separately asynchronously, the pseudocode as follows:
```
Task A;
Executor().run(A);
Task B;
B.succed(A);
Executor().run(B);
Task C;
C.succed(A);
Executor().run(C);
Task D;
D.succed(B, C)
Executor().run(D);
```
The main program should dynamically generate tasks and execute them concurrently using executor with considering their dependencies. Our question is whether taskflow supports the above scenario. ps: From reading the example and handbook, as far as we know, the entire graph is executed directly if executor.run(taskflow) is invoked, while executor.async() does not consider dependencies between tasks.

**Comment by tsung-wei-huang** [comment]
@ccsuzzh currently, we do not support this feature - but we are working on an interface exactly similar to what you mention here. Hopefully it will come up in the next release.

**Comment by ccsuzzh** [comment]
Thanks for your answer, I'm very much looking forward to the next release.

**Comment by tsung-wei-huang** [comment]
@ccsuzzh We have the first version of dynamic task graph parallelism using [dependent asynchronous tasking](https://taskflow.github.io/taskflow/DependentAsyncTasking.html). It is being experimented in both master and dev branches and will be included in the next release 3.6.0. 

Please give it a try :)

**Comment by ccsuzzh** [comment]
Thanks for the quick reply! 
I've also been experimenting with dependent asynchronous tasks recently, and it meets previous expectations very well. Sorry for the late reply. 
Close this issue now.
